File: css/cssInterface.py
#!/usr/bin/env python

# LSST Data Management System
# Copyright 2013 LSST Corporation.
# 
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the LSST License Statement and 
# the GNU General Public License along with this program.  If not, 
# see <http://www.lsstcorp.org/LegalNotices/>.


"""
This module implements interface to the Central State System (CSS).

@author  Jacek Becla, SLAC


Known issues and todos:
 - connection information for the Kazoo Client needs to be configurable.
 - recover from lost connection by reconnecting
 - issue: watcher is currently using the "_zk", and bypasses the official API!
 - implement proper logging instead of print

"""

import time

from kazoo.client import KazooClient


####################################################################################
####################################################################################
####################################################################################
class CssException(Exception):
    """
    Exception raised by CSSInterface.
    """

    SUCCESS                     =    0
    ERR_DB_DOES_NOT_EXIST       = 2001
    ERR_KEY_ALREADY_EXISTS      = 2002
    ERR_KEY_DOES_NOT_EXIST      = 2003
    ERR_KEY_INVALID             = 2004
    ERR_NOT_IMPLEMENTED         = 9998
    ERR_INTERNAL                = 9999

    ################################################################################
    def __init__(self, errNo, extraMsgList=None):
        """
        Initialize the shared data.

        @param errNo      Error number.
        @param extraMsgList  Optional list of extra messages.
        """
        self._errNo = errNo
        self._extraMsgList = extraMsgList

        self._errors = { 
            CssException.ERR_DB_DOES_NOT_EXIST: ("Database does not exist."),
            CssException.ERR_KEY_ALREADY_EXISTS: ("Key already exists."),
            CssException.ERR_KEY_INVALID: ("Invalid key."),
            CssException.ERR_KEY_DOES_NOT_EXIST: ("Key does not exist."),
            CssException.ERR_NOT_IMPLEMENTED: ("Fature not implemented yet."),
            CssException.ERR_INTERNAL: "Internal error."
        }

    ################################################################################
    def __str__(self):
        """
        Return string representation of the error.

        @return string  Error message string, including all optional messages.
        """
        msg = self._errors.get(self._errNo, "Undefined css error.")
        if self._extraMsgList is not None: 
            for s in self._extraMsgList: msg += " (%s)" % s
        return msg

    ################################################################################
    def getErrNo(self):
        """
        Return error number.
        """
        return self._errNo

####################################################################################
####################################################################################
####################################################################################
class CssInterface(object):
    """
    @brief CssInterface class defines interface to the Central State Service CSS).
    """

    ################################################################################
    def __init__(self, verbose=True):
        """
        Initialize the interface.

        @param verbose Verbose flag, default is True.
        """
        self._zk = KazooClient(hosts='127.0.0.1:2181') # FIXME
        self._zk.start()
        self._verbose = verbose

    ################################################################################
    def create(self, k, v='', sequence=False):
        """
        Add a new key/value entry. Create entire path as necessary. 

        @param sequence  Sequence flag -- if set to True, a 10-digid, 0-padded
                         suffix (unique sequential number) will be added to the key.
        @return string   Real path to the just created node.
        """
        # check if the key exists
        if self._zk.exists(k):
            raise CssException(CssException.ERR_KEY_ALREADY_EXISTS, [k])
        p = self._chopLastSection(k)
        if p is None:
            raise CssException(CssException.ERR_KEY_INVALID, [k])
        if self._verbose: print "cssInterface: CREATE '%s' --> '%s'" % (k, v) 
        return self._zk.create(k, v, sequence=sequence, makepath=True)

    ################################################################################
    def exists(self, k):
        """
        Check if a given key exists.

        @param k Key.

        @return boolean  True if the key exists, False otherwise.
        """
        return self._zk.exists(k)

    ################################################################################
    def get(self, k):
        """
        Get value for a key. Raise exception if the key doesn't exist.

        @param k   Key.

        @return string  Value for a given key. 
        """
        if not self._zk.exists(k):
            raise CssException(CssException.ERR_KEY_DOES_NOT_EXIST, [k])
        v, stat = self._zk.get(k)
        if self._verbose: print "cssInterface: GET '%s' --> '%s'" % (k, v)
        return v

    ################################################################################
    def getChildren(self, k):
        """
        Get children for a given key. Raise exception if the key doesn't exist.

        @param k   Key.

        @return    List_of_strings  A list of children for a given key. 
        """
        if not self._zk.exists(k):
            raise CssException(CssException.ERR_KEY_DOES_NOT_EXIST, [k])
        if self._verbose: print "cssInterface: GETCHILDREN '%s'" % (k)
        return self._zk.get_children(k)

    ################################################################################
    def set(self, k, v):
        """
        Set value for a given key. Raise exception if the key doesn't exist.

        @param k  Key.
        @param v  Value.
        """
        # check if the key exists
        if not self._zk.exists(k):
            raise CssException(CssException.ERR_KEY_DOES_NOT_EXIST, [k])
        v1, stat = self._zk.get(k)
        if self._verbose: print "cssInterface: SET '%s' --> '%s'" % (k, v)
        self._zk.set(k, v)
        v2, stat = self._zk.get(k)

    ################################################################################
    def delete(self, k, recursive=False):
        """
        Delete a key, including all children if recursive flag is set.

        @param k         Key.
        @param recursive Flag. If set, all existing children nodes will be
                         deleted.

        Delete a key, including all children if recursive flag is set. Raise
        exception if the key doesn't exist.
        """
        if not self._zk.exists(k):
            raise CssException(CssException.ERR_KEY_DOES_NOT_EXIST, [k])
        if self._verbose: print "cssInterface: DELETE '%s'" % (k)
        self._zk.delete(k, recursive=recursive)

    ################################################################################
    def deleteAll(self, p):
        """
        Delete everything recursively starting from a given point in the tree.

        @param p  Path.
        """
        if self._zk.exists(p):
            self._deleteOne(p)

    ################################################################################
    def printAll(self):
        """
        Print entire contents to stdout.
        """
        self._printOne("/")

    ################################################################################
    def startTransaction(self):
        """
        Start transaction and return transactionRequest instance.
        """
        return self._zk.transaction()

    ################################################################################
    def _chopLastSection(self, k):
        """
        Remove substring after last '/', e.g. for /xx/y/abc it'll return /xx/y.

        @param k  Key.

        @return string
        """
        x = k.rfind('/')
        if x == -1: return None
        return k[0:x]

    ################################################################################
    def _printOne(self, p):
        """
        Print content of one znode. Note, this function is recursive.

        @param p  Path.

        @return   string
        """
        t = self.startTransaction()
        children = None
        data = None
        stat = None
        if self.exists(p):
            children = self._zk.get_children(p)
            data, stat = self._zk.get(p)
        t.commit()

        print p, "=", data
        for child in children:
            if p == "/":
                if child != "zookeeper":
                    self._printOne("%s%s" % (p, child))
            else:
                self._printOne("%s/%s" % (p, child))

    ################################################################################
    def _deleteOne(self, p):
        """
        Delete one znode. Note, this function is recursive.

        @param p  Path.
        """
        children = self._zk.get_children(p)
        for child in children:
            if p == "/":
                if child != "zookeeper": # skip "/zookeeper"
                    self._deleteOne("%s%s" % (p, child))
            else:
                self._deleteOne("%s/%s" % (p, child))
        if p != "/": 
            self._zk.delete(p)

File: css/cssInterface.cc
/*
 * LSST Data Management System
 * Copyright 2014 LSST Corporation.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */

/**
  * @file CssInterface.cc
  *
  * @brief Interface to the Common State System.
  *
  * @Author Jacek Becla, SLAC
  */

/*
 * Based on:
 * http://zookeeper.apache.org/doc/r3.3.4/zookeeperProgrammers.html#ZooKeeper+C+client+API
 *
 * To do:
 *  - logging
 *  - perhaps switch to async (seems to be recommended by zookeeper)
 */


// standard library imports
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string.h> // for memset

// local imports
#include "cssInterface.h"


using std::cout;
using std::endl;
using std::exception;
using std::ostringstream;
using std::string;
using std::vector;

namespace qCss = lsst::qserv::css;

qCss::CssInterface::CssInterface(string const& connInfo) {
    zoo_set_debug_level(ZOO_LOG_LEVEL_DEBUG);
    _zh = zookeeper_init(connInfo.c_str(), 0, 10000, 0, 0, 0);
    if ( !_zh ) {
        throw std::runtime_error("Failed to connect");
    }
}

qCss::CssInterface::~CssInterface() {
    zookeeper_close(_zh);
}

void
qCss::CssInterface::create(string const& key, string const& value) {
    cout << "*** CssInterface::create(), " << key << " --> " << value << endl;
    char buffer[512];
    int rc = zoo_create(_zh, key.c_str(), value.c_str(), value.length(), 
                        &ZOO_OPEN_ACL_UNSAFE, ZOO_EPHEMERAL,
                        buffer, sizeof(buffer)-1);
    if ( rc ) {
        ostringstream s;
        s << "zoo_create failed, error: " << rc;
        throw std::runtime_error(s.str());
    }
}

bool
qCss::CssInterface::exists(string const& key) {
    cout << "*** CssInterface::exist(), key: " << key << endl;
    struct Stat stat;
    int rc = zoo_exists(_zh, key.c_str(), 0,  &stat);
    return rc == 0;
}

string
qCss::CssInterface::get(string const& key) {
    cout << "*** CssInterface::get(), key: " << key << endl;
    char buffer[512];
    memset(buffer, 0, 512);
    int buflen = sizeof(buffer);
    struct Stat stat;
    int rc = zoo_get(_zh, key.c_str(), 0, buffer, &buflen, &stat);
    if ( rc ) {
        ostringstream s;
        s << "zoo_get failed, error: " << rc;
        throw std::runtime_error(s.str());
    }
    cout << "*** got " << buffer << endl;
    return string(buffer);
}

std::vector<string> 
qCss::CssInterface::getChildren(string const& key) {
    cout << "*** CssInterface::getChildren, key: " << key << endl;
    struct String_vector strings;
    int rc = zoo_get_children(_zh, key.c_str(), 0, &strings);
    if ( rc ) {
        throw std::runtime_error("Key not found");
    }
    cout << "got " << strings.count << " children" << endl;
    vector<string> v;
    int i;
    for (i=0 ; i < strings.count ; i++) {
        cout << "   " << i+1 << ": " << strings.data[i] << endl;
        v.push_back(strings.data[i]);
    }
    return v;
}

void
qCss::CssInterface::deleteNode(string const& key) {
    int rc = zoo_delete(_zh, key.c_str(), -1);
}
